esphome/components/emporia_vue/emporia_vue.cpp

```diff
@@ -8,114 +9,131 @@ namespace emporia_vue {
+                 sensor_reading.sequence_num - last_sequence_num - 1);
+      }
+
+      xQueueOverwrite(global_emporia_vue_component->i2c_data_queue_, &sensor_reading);
+      ESP_LOGV(TAG, "Added sensor reading with sequence number %d to queue", sensor_reading.sequence_num);
+
+      last_sequence_num = sensor_reading.sequence_num;
+      vTaskDelayUntil(&last_poll, poll_interval);
     }
-
-    vTaskDelayUntil(&xLastWakeTime, xDelay);
   }
 }
 
 void EmporiaVueComponent::loop() {
-  EmporiaSensorData data;
+  SensorReading sensor_reading;
 
-  if (xQueueReceive(this->i2c_data_queue_, &data, 0 ) == pdTRUE) {
+  if (xQueueReceive(this->i2c_data_queue_, &sensor_reading, 0) == pdTRUE) {
+    ESP_LOGV(TAG, "Received sensor reading with sequence number %d from queue", sensor_reading.sequence_num);
     for (PhaseConfig *phase : this->phases_) {
-      phase->update_from_data(data);
+      phase->update(sensor_reading);
     }
 
     for (CTSensor *ct_sensor : this->ct_sensors_) {
-      ct_sensor->update_from_data(data);
+      ct_sensor->update(sensor_reading);
     }
   }
 }
 
-void PhaseConfig::update_from_data(EmporiaSensorData &data) {
-  float calibrated_voltage = data.voltage[this->input_wire_] * this->calibration_;
-  this->voltage_sensor_.publish_state(calibrated_voltage);
+void PhaseConfig::update(const SensorReading &sensor_reading) {
+  if (this->voltage_sensor_) {
+    float calibrated_voltage = sensor_reading.voltage[this->input_wire_] * this->calibration_;
+    this->voltage_sensor_->publish_state(calibrated_voltage);
+  }
 }
 
-int32_t PhaseConfig::extract_power_for_phase(const PowerDataEntry &entry) {
+int32_t PhaseConfig::extract_power_for_phase(const ReadingPowerEntry &power_entry) {
   switch (this->input_wire_) {
-    case PhaseInputColor::BLACK:
-      return entry.phase_one;
-    case PhaseInputColor::RED:
-      return entry.phase_two;
-    case PhaseInputColor::BLUE:
-      return entry.phase_three;
+    case PhaseInputWire::BLACK:
+      return power_entry.phase_black;
+    case PhaseInputWire::RED:
+      return power_entry.phase_red;
+    case PhaseInputWire::BLUE:
+      return power_entry.phase_blue;
     default:
+      ESP_LOGE(TAG, "Unsupported phase input wire, this should never happen");
       return -1;
   }
 }
 
-void CTSensor::update_from_data(const EmporiaSensorData &data) {
-  PowerDataEntry entry = data.power[this->ct_input_];
-  int32_t raw_power = this->phase_->extract_power_for_phase(entry);
+void CTSensor::update(const SensorReading &sensor_reading) {
+  ReadingPowerEntry power_entry = sensor_reading.power[this->input_port_];
+  int32_t raw_power = this->phase_->extract_power_for_phase(power_entry);
   float calibrated_power = this->get_calibrated_power(raw_power);
   this->publish_state(calibrated_power);
 }
 
-double CTSensor::get_calibrated_power(int32_t raw_power) {
-  double calibration = this->phase_->get_calibration();
+float CTSensor::get_calibrated_power(int32_t raw_power) const {
```

Comment:
I'm not sure why the `const` is needed here?

Reply:
It's not needed, just a nice to have; it signals that this function has no side effects on the members of the class, so making any changes to the class from this method would cause a compile error. I think it's best practice to set it to `const` when possible

Reply:
Got it - been a while since I've done C - C++ so I wasn't exactly sure, thanks